#  Solve the problem

## Introduction

What is a Regression Problem? A Regression Problem is when an output/target variable is real or continuous value. Regression is supervised machine learning technique which is used in prediction of these continuous variables such as height, salary & house price😉 too.

So what exactly we are trying to do? We are trying to : Predict a Target Using some features

Now let's come to the competion problem statement


## Challenge Objectives
In context of the details shared above via the Project Background, in this challenge we are looking to introduce a straightforward regression practice problem.

The dataset has been shared in the challenge forum. There are 45 features in the dataset, and one regression label (i.e. it's a single-label prediction challenge, not multi-label)

Within this goal the following are the targets:

Train a model, using the file train.csv - The train.csv file can be found in the challenge dataset folder in the forum. That file should be used to train a machine learning model. Here the last (right-most) column in the file is the label, and all other columns are features
Using the trained model, predict regression labels for test_x.csv - The test_x.csv file contains the test dataset. It does NOT contain the labels, and hence the regression labels need to be predicted.

## Importing necessary libraries📚

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
np.random.seed(123)
warnings.filterwarnings('ignore')
%matplotlib inline

## Load the train and test dataset.

In [2]:
# Read training and testing csv file
# Let's see what's inside🧐

# Import data
train = pd.read_csv('train.csv',header=None)
test = pd.read_csv('test_x.csv',header=None)

## Let’s observe the shape of our datasets.

In [3]:
# print shape
print('train data shape :', train.shape)
print('test data shape :', test.shape)

train data shape : (14700, 46)
test data shape : (6300, 45)


In [4]:
# inspect train data
train.head()

0            1            2            3            4   \
0 -4660.197737  2665.495526 -1104.915343   563.350855     6.945110   
1  5290.777109 -8871.620989  2040.742184  1941.290147  -278.315345   
2 -5079.817347   246.732978   -82.682441   358.441882   639.135525   
3 -3179.838400   129.635078  1198.864399  -404.091302  1017.011710   
4  8031.785309 -3133.016055 -1792.306828  1043.586308  -304.023415   

           5           6           7           8           9   ...        36  \
0  273.731752  257.039238 -127.932947   37.314456  -46.378287  ...  2.103955   
1    8.565463  -17.412734  348.533407  371.050559   58.740439  ...  2.858535   
2   52.071644  562.167586 -281.638695   78.059886   18.453052  ... -7.614588   
3  396.553775 -326.369718  604.641433  -28.455874 -178.166313  ...  0.181635   
4 -216.729002  250.475959  139.402207 -118.387878   86.607414  ... -4.013917   

         37        38        39        40        41        42        43  \
0  2.560691 -2.061088  1.112229  5.690221  3.659967 -0.706192  4.152445   
1 -4.209751  2.743391  1.083289  8.776751 -2.481668  1.106314 -0.053182   
2  0.007990  0.232106 -1.647424 -2.217109 -3.508988 -1.078488  1.619875   
3 -3.578278 -3.703888  0.259553  0.377226 -2.955690  0.658655 -3.193840   
4  1.697319  2.200147 -5.272229  3.394900  0.510452  0.825802 -3.106947   

         44      45  
0 -1.941093  108.00  
1  3.214643    2.56  
2 -2.545034   34.00  
3 -1.492794   26.00  
4  2.247548    4.90  

[5 rows x 46 columns]

In [5]:
# inspect test dataset

test.head()

0             1            2            3           4   \
0  2515.936305   7288.285512 -1858.830138   222.574033 -916.797456   
1  2454.139714    103.556679  -598.520955   462.366973 -335.328987   
2  5609.467216   2157.838672 -2314.190682  -170.946082   13.434814   
3  3664.016902 -10135.360948 -2513.506198 -1210.826651  370.625464   
4 -4537.429816     61.474415  -586.041802   633.223813  736.243819   

            5           6           7           8           9   ...        35  \
0 -1047.550169 -127.957684  195.880387  -60.634930  273.436423  ... -3.122042   
1 -1095.038558  -85.263061  100.783570  -92.923532   13.724276  ...  5.194352   
2 -1000.952134 -377.684223  230.026122  127.914059   50.447168  ...  2.394519   
3   339.078681  105.921730  203.945078  449.974420   45.771650  ... -5.844049   
4   -55.852641 -208.141954 -207.778268   93.310093   51.814536  ...  0.219893   

         36        37        38        39        40        41        42  \
0  3.512524  5.186542 -2.442969 -1.368228 -3.856474 -4.693315 -4.375151   
1 -7.053990  7.786843 -2.567848  1.901989 -1.818009  1.174387  0.314798   
2 -4.873193 -0.648127 -0.058641  2.414594  5.977978 -6.973337  0.142927   
3 -0.564908 -6.202512  7.311372 -2.192002  1.708209 -0.913340  0.255181   
4  2.074770  3.863877  0.367331  1.889209 -1.620242  1.263507  3.439319   

         43        44  
0 -5.174799 -0.670157  
1 -1.834345  1.063272  
2  0.145413  4.528342  
3  3.767441  6.589483  
4  0.219539 -1.755496  

[5 rows x 45 columns]

In [6]:
train.dtypes

0     float64
1     float64
2     float64
3     float64
4     float64
5     float64
6     float64
7     float64
8     float64
9     float64
10    float64
11    float64
12    float64
13    float64
14    float64
15    float64
16    float64
17    float64
18    float64
19    float64
20    float64
21    float64
22    float64
23    float64
24    float64
25    float64
26    float64
27    float64
28    float64
29    float64
30    float64
31    float64
32    float64
33    float64
34    float64
35    float64
36    float64
37    float64
38    float64
39    float64
40    float64
41    float64
42    float64
43    float64
44    float64
45    float64
dtype: object

### Examine Missing Values

In [7]:
train.isnull().sum().sum()

0

In [8]:
test.isnull().sum().sum()

0

In [9]:
train.describe().T

count       mean          std           min          25%          50%  \
0   14700.0  -2.928685  6872.953260  -8068.771503 -4767.271846 -3451.622656   
1   14700.0 -17.518314  4784.468064 -11211.410660 -2275.025951     3.888990   
2   14700.0  11.459230  1933.993280 -10205.002558  -755.818656    -8.958387   
3   14700.0  -8.162813   890.619064  -3684.714678  -596.777993    74.874144   
4   14700.0   0.655671   811.539319  -4534.040750  -547.710341    93.983217   
5   14700.0  -4.043439   633.328574  -2305.025359  -271.734816   -83.399018   
6   14700.0  -0.138544   336.898341  -1730.204557  -179.435951    -2.845638   
7   14700.0  -0.006607   245.082789  -1566.521099  -150.598860   -47.721963   
8   14700.0   0.066400   182.482671   -783.203839   -79.440460     6.842191   
9   14700.0   0.556442   126.622677   -505.731353   -72.675584    -7.721030   
10  14700.0  -0.340992   107.529987   -482.940947   -52.271028    -0.030772   
11  14700.0  -0.177790   100.919051   -358.963908   -45.992272    -2.450377   
12  14700.0   0.138595    55.939989   -239.929493   -27.965722     5.445070   
13  14700.0  -0.014887    52.083453   -301.953558   -30.499770    -4.123659   
14  14700.0   0.037044    47.219189   -291.080237   -24.056559     2.005840   
15  14700.0  -0.003637    38.176517   -208.267902   -21.888210    -1.762408   
16  14700.0   0.073480    36.715329   -162.084689   -20.346971     0.399135   
17  14700.0   0.154437    32.430475   -173.997418   -19.480783    -3.051493   
18  14700.0  -0.022231    30.170541   -142.821054   -15.617664    -2.342461   
19  14700.0  -0.216096    25.115337   -139.801099   -12.470553    -0.506555   
20  14700.0  -0.005635    21.255090    -82.291652    -9.820247    -0.648865   
21  14700.0   0.003599    20.508035    -99.411265   -11.452188     1.019940   
22  14700.0   0.075826    16.664873    -86.803102    -9.260807    -0.643919   
23  14700.0   0.265845    16.034779    -75.148941   -10.165276    -0.947252   
24  14700.0  -0.028508    15.231448    -74.532016   -10.024967     0.041602   
25  14700.0  -0.095273    13.261767   -115.218538    -7.220194     0.329099   
26  14700.0   0.019302    12.319667    -64.551403    -6.619701    -0.020191   
27  14700.0  -0.022060    11.013524    -54.542476    -5.624598    -0.187433   
28  14700.0   0.014246     9.815995    -53.822157    -4.800634    -0.358477   
29  14700.0   0.075194     9.216669    -49.570500    -5.904047    -0.306743   
30  14700.0  -0.031862     7.709913    -49.732670    -4.805265    -0.197511   
31  14700.0  -0.000355     7.529075    -49.979342    -3.541031     0.674311   
32  14700.0  -0.004274     6.171812    -31.940791    -3.398998    -0.179688   
33  14700.0  -0.016141     5.920547    -28.736378    -3.039510    -0.344400   
34  14700.0  -0.004034     5.566140    -55.434389    -2.539631    -0.296778   
35  14700.0   0.010824     4.972793    -32.443042    -2.723814    -0.188577   
36  14700.0   0.025553     4.578815    -25.312110    -2.808594    -0.029756   
37  14700.0  -0.006114     4.150619    -25.453918    -2.089203     0.051858   
38  14700.0   0.015436     3.931963    -19.517281    -2.383933    -0.282553   
39  14700.0  -0.006860     3.556961    -24.861145    -1.796503     0.293436   
40  14700.0   0.002940     3.476686    -20.426642    -1.926005    -0.120790   
41  14700.0   0.010346     3.056496    -13.369509    -1.805738     0.091836   
42  14700.0   0.024842     2.935054    -17.393424    -1.501565     0.036378   
43  14700.0  -0.009424     2.696886    -19.555663    -1.591546    -0.169429   
44  14700.0  -0.013992     2.389618    -14.484096    -1.273894    -0.064081   
45  14700.0  34.323663    34.217109      0.000500     5.360000    20.000000   

            75%           max  
0   3811.057396  39329.032795  
1   1297.606081  18250.700651  
2    700.310862  14322.991245  
3    532.066305   7861.934157  
4    464.185555   5451.325961  
5    277.676436   4465.064803  
6    177.507713   2667.573030  
7    131.271410   2767.293212  
8   

# Model Development

This section includes below steps:

Data preparation in order to feed the model

Develop different models and compare them.

In [10]:
# importing python libaries

import time
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import ElasticNet
from sklearn.linear_model import SGDRegressor
from sklearn.svm import SVR
from  sklearn.linear_model import BayesianRidge
from catboost import CatBoostRegressor
from sklearn.kernel_ridge import KernelRidge
from sklearn.linear_model import LinearRegression
from xgboost.sklearn import XGBRegressor
from lightgbm import  LGBMRegressor


from sklearn.metrics import mean_absolute_error, mean_squared_error,mean_absolute_error,r2_score,accuracy_score
from sklearn.model_selection import train_test_split

In [11]:
# splitting dataset in train abf val
X=train.iloc[:,0: -1]
Y = train.iloc[:, -1]
test_size=0.20
seed=42
X_train, X_Val, Y_train, Y_Val = train_test_split(X, Y, test_size=test_size,random_state=seed)

In [12]:
def Models(model, name, d):
    print("Working on {} model".format(name))
    
    
    cla = model
    cla.fit(X_train, Y_train)
    
   
    
    predicted = cla.predict(X_Val)
   
    
    MAE=mean_absolute_error(Y_Val,predicted)
    MSE=mean_squared_error(Y_Val,predicted)
    RMSE=mean_squared_error(Y_Val,predicted,squared=False)
    MAEA=mean_absolute_error(Y_Val,predicted)
    R2=r2_score(Y_Val,predicted)
    
    
    
    d['Name'].append(name)
    #d['Training ACU'].append(tr_auc)
    #d['Testing ACU'].append(te_auc)
    d["RMSE"].append(RMSE)
    d['MAE'].append(MAE)
    d['MSE'].append(MSE)
    d['MAEA'].append(MAEA)
    d['R2'].append(R2)
    
    
    
    
    print("**********"*5)
    print()
    return d
    
    
    
    
     

In [13]:

d = {'Name' : [], 'RMSE': [], 'R2': [], 'MAE': [], 'MSE': [],'MAEA': []}


models = [ [GradientBoostingRegressor(loss="squared_error",learning_rate=0.2,n_estimators=1000), 'GradientBoostRegressor'],
          [ElasticNet(max_iter=2000,normalize=True), 'ElasticNet'], 
          [LGBMRegressor(), 'LightGBM'],
          [SGDRegressor(l1_ratio=0.5,), "SGDRegressor"],

          [CatBoostRegressor(), 'CatBoost'], 
          [KernelRidge(alpha=1.0), 'kernal Ridge'],
          [LinearRegression(), 'Linear Regression'], 
          [SVR(C=1.0,epsilon=0.2), 'Support Vector Ratio' ],
         [XGBRegressor(objective="reg:linear"),"XGBRegressor"],
         [BayesianRidge(),"BayesianRidge"]
         ]




for model in models:
    d = Models(model[0], model[1], d)

acu_data = pd.DataFrame(data = d)

Working on GradientBoostRegressor model
**************************************************

Working on ElasticNet model
**************************************************

Working on LightGBM model
**************************************************

Working on SGDRegressor model
**************************************************

Working on CatBoost model
Learning rate set to 0.060437
0:	learn: 32.9024384	total: 73.3ms	remaining: 1m 13s
1:	learn: 31.6983334	total: 91.6ms	remaining: 45.7s
2:	learn: 30.5833273	total: 111ms	remaining: 36.8s
3:	learn: 29.5147468	total: 129ms	remaining: 32.2s
4:	learn: 28.5525952	total: 148ms	remaining: 29.4s
5:	learn: 27.6334746	total: 174ms	remaining: 28.8s
6:	learn: 26.7772731	total: 201ms	remaining: 28.6s
7:	learn: 25.9933962	total: 224ms	remaining: 27.8s
8:	learn: 25.3138427	total: 248ms	remaining: 27.3s
9:	learn: 24.6846390	total: 273ms	remaining: 27s
10:	learn: 24.0696899	total: 294ms	remaining: 26.4s
11:	learn: 23.4832847	total: 311ms	remaining: 25.

161:	learn: 12.8676415	total: 3.54s	remaining: 18.3s
162:	learn: 12.8501772	total: 3.56s	remaining: 18.3s
163:	learn: 12.8322061	total: 3.58s	remaining: 18.3s
164:	learn: 12.8176923	total: 3.6s	remaining: 18.2s
165:	learn: 12.8031355	total: 3.62s	remaining: 18.2s
166:	learn: 12.7678716	total: 3.64s	remaining: 18.2s
167:	learn: 12.7518088	total: 3.66s	remaining: 18.1s
168:	learn: 12.7416376	total: 3.68s	remaining: 18.1s
169:	learn: 12.7249848	total: 3.7s	remaining: 18s
170:	learn: 12.7087465	total: 3.71s	remaining: 18s
171:	learn: 12.6885747	total: 3.73s	remaining: 18s
172:	learn: 12.6719506	total: 3.75s	remaining: 17.9s
173:	learn: 12.6609892	total: 3.78s	remaining: 17.9s
174:	learn: 12.6432268	total: 3.8s	remaining: 17.9s
175:	learn: 12.6185117	total: 3.81s	remaining: 17.9s
176:	learn: 12.6011779	total: 3.83s	remaining: 17.8s
177:	learn: 12.5880441	total: 3.85s	remaining: 17.8s
178:	learn: 12.5769117	total: 3.87s	remaining: 17.8s
179:	learn: 12.5622848	total: 3.89s	remaining: 17.7s
18

319:	learn: 10.8342655	total: 6.62s	remaining: 14.1s
320:	learn: 10.8160138	total: 6.65s	remaining: 14.1s
321:	learn: 10.8049304	total: 6.67s	remaining: 14.1s
322:	learn: 10.8007421	total: 6.7s	remaining: 14s
323:	learn: 10.7913900	total: 6.71s	remaining: 14s
324:	learn: 10.7812807	total: 6.73s	remaining: 14s
325:	learn: 10.7757986	total: 6.75s	remaining: 14s
326:	learn: 10.7671391	total: 6.77s	remaining: 13.9s
327:	learn: 10.7597480	total: 6.79s	remaining: 13.9s
328:	learn: 10.7544768	total: 6.8s	remaining: 13.9s
329:	learn: 10.7468660	total: 6.83s	remaining: 13.9s
330:	learn: 10.7383233	total: 6.85s	remaining: 13.8s
331:	learn: 10.7293120	total: 6.87s	remaining: 13.8s
332:	learn: 10.7250116	total: 6.88s	remaining: 13.8s
333:	learn: 10.7132236	total: 6.9s	remaining: 13.8s
334:	learn: 10.7030921	total: 6.92s	remaining: 13.7s
335:	learn: 10.6987993	total: 6.94s	remaining: 13.7s
336:	learn: 10.6829397	total: 6.96s	remaining: 13.7s
337:	learn: 10.6738430	total: 6.98s	remaining: 13.7s
338:

481:	learn: 9.6869876	total: 9.77s	remaining: 10.5s
482:	learn: 9.6814359	total: 9.79s	remaining: 10.5s
483:	learn: 9.6725668	total: 9.82s	remaining: 10.5s
484:	learn: 9.6656063	total: 9.84s	remaining: 10.4s
485:	learn: 9.6596676	total: 9.86s	remaining: 10.4s
486:	learn: 9.6528026	total: 9.88s	remaining: 10.4s
487:	learn: 9.6461832	total: 9.89s	remaining: 10.4s
488:	learn: 9.6432308	total: 9.91s	remaining: 10.4s
489:	learn: 9.6370715	total: 9.93s	remaining: 10.3s
490:	learn: 9.6321163	total: 9.95s	remaining: 10.3s
491:	learn: 9.6229689	total: 9.97s	remaining: 10.3s
492:	learn: 9.6160409	total: 9.99s	remaining: 10.3s
493:	learn: 9.6068496	total: 10s	remaining: 10.3s
494:	learn: 9.5987889	total: 10s	remaining: 10.2s
495:	learn: 9.5955901	total: 10s	remaining: 10.2s
496:	learn: 9.5891087	total: 10.1s	remaining: 10.2s
497:	learn: 9.5826243	total: 10.1s	remaining: 10.2s
498:	learn: 9.5779972	total: 10.1s	remaining: 10.1s
499:	learn: 9.5732286	total: 10.1s	remaining: 10.1s
500:	learn: 9.5686

640:	learn: 8.8911682	total: 12.8s	remaining: 7.2s
641:	learn: 8.8865559	total: 12.9s	remaining: 7.18s
642:	learn: 8.8831649	total: 12.9s	remaining: 7.16s
643:	learn: 8.8786933	total: 12.9s	remaining: 7.14s
644:	learn: 8.8748853	total: 12.9s	remaining: 7.12s
645:	learn: 8.8713846	total: 13s	remaining: 7.1s
646:	learn: 8.8657968	total: 13s	remaining: 7.08s
647:	learn: 8.8619130	total: 13s	remaining: 7.06s
648:	learn: 8.8570806	total: 13s	remaining: 7.04s
649:	learn: 8.8538430	total: 13s	remaining: 7.02s
650:	learn: 8.8506426	total: 13.1s	remaining: 7s
651:	learn: 8.8474201	total: 13.1s	remaining: 6.98s
652:	learn: 8.8433799	total: 13.1s	remaining: 6.96s
653:	learn: 8.8386793	total: 13.1s	remaining: 6.94s
654:	learn: 8.8358227	total: 13.1s	remaining: 6.92s
655:	learn: 8.8305683	total: 13.2s	remaining: 6.9s
656:	learn: 8.8275065	total: 13.2s	remaining: 6.88s
657:	learn: 8.8260202	total: 13.2s	remaining: 6.85s
658:	learn: 8.8213024	total: 13.2s	remaining: 6.83s
659:	learn: 8.8168084	total:

802:	learn: 8.3020659	total: 16s	remaining: 3.93s
803:	learn: 8.2990822	total: 16s	remaining: 3.91s
804:	learn: 8.2967165	total: 16.1s	remaining: 3.89s
805:	learn: 8.2911642	total: 16.1s	remaining: 3.87s
806:	learn: 8.2893176	total: 16.1s	remaining: 3.85s
807:	learn: 8.2864352	total: 16.1s	remaining: 3.83s
808:	learn: 8.2822819	total: 16.1s	remaining: 3.81s
809:	learn: 8.2802319	total: 16.2s	remaining: 3.79s
810:	learn: 8.2779649	total: 16.2s	remaining: 3.77s
811:	learn: 8.2754832	total: 16.2s	remaining: 3.75s
812:	learn: 8.2723099	total: 16.2s	remaining: 3.73s
813:	learn: 8.2667561	total: 16.2s	remaining: 3.71s
814:	learn: 8.2625044	total: 16.3s	remaining: 3.69s
815:	learn: 8.2611201	total: 16.3s	remaining: 3.67s
816:	learn: 8.2570173	total: 16.3s	remaining: 3.65s
817:	learn: 8.2558741	total: 16.3s	remaining: 3.63s
818:	learn: 8.2528934	total: 16.3s	remaining: 3.61s
819:	learn: 8.2504403	total: 16.3s	remaining: 3.59s
820:	learn: 8.2491822	total: 16.4s	remaining: 3.57s
821:	learn: 8.24

961:	learn: 7.8387868	total: 19.1s	remaining: 755ms
962:	learn: 7.8357784	total: 19.1s	remaining: 735ms
963:	learn: 7.8334849	total: 19.2s	remaining: 716ms
964:	learn: 7.8294226	total: 19.2s	remaining: 696ms
965:	learn: 7.8263465	total: 19.2s	remaining: 676ms
966:	learn: 7.8233048	total: 19.2s	remaining: 656ms
967:	learn: 7.8206503	total: 19.2s	remaining: 636ms
968:	learn: 7.8186910	total: 19.3s	remaining: 616ms
969:	learn: 7.8179403	total: 19.3s	remaining: 596ms
970:	learn: 7.8150690	total: 19.3s	remaining: 576ms
971:	learn: 7.8139032	total: 19.3s	remaining: 556ms
972:	learn: 7.8110152	total: 19.3s	remaining: 536ms
973:	learn: 7.8087056	total: 19.4s	remaining: 517ms
974:	learn: 7.8072923	total: 19.4s	remaining: 497ms
975:	learn: 7.8056464	total: 19.4s	remaining: 477ms
976:	learn: 7.8020071	total: 19.4s	remaining: 457ms
977:	learn: 7.8002087	total: 19.4s	remaining: 437ms
978:	learn: 7.7979510	total: 19.4s	remaining: 417ms
979:	learn: 7.7946992	total: 19.5s	remaining: 397ms
980:	learn: 

In [14]:
acu_data

Name          RMSE            R2           MAE  \
0  GradientBoostRegressor  1.211325e+01  8.770419e-01  7.744536e+00   
1              ElasticNet  3.454650e+01 -1.014129e-04  2.944331e+01   
2                LightGBM  1.179314e+01  8.834546e-01  7.530351e+00   
3            SGDRegressor  6.530468e+16 -3.573753e+30  5.075457e+16   
4                CatBoost  1.145858e+01  8.899734e-01  7.149985e+00   
5            kernal Ridge  3.960898e+01 -3.146897e-01  3.553333e+01   
6       Linear Regression  1.916455e+01  6.922250e-01  1.470786e+01   
7    Support Vector Ratio  2.376872e+01  5.265792e-01  1.745765e+01   
8            XGBRegressor  1.171729e+01  8.849491e-01  7.169678e+00   
9           BayesianRidge  1.915884e+01  6.924086e-01  1.470827e+01   

            MSE          MAEA  
0  1.467307e+02  7.744536e+00  
1  1.193461e+03  2.944331e+01  
2  1.390783e+02  7.530351e+00  
3  4.264702e+33  5.075457e+16  
4  1.312992e+02  7.149985e+00  
5  1.568872e+03  3.553333e+01  
6  3.672801e+02  1.470786e+01  
7  5.649518e+02  1.745765e+01  
8  1.372948e+02  7.169678e+00  
9  3.670611e+02  1.470827e+01

In [15]:
# Hyperparameter Optimization
xgb_parms={"max_depth":[3,4,5,6,10,12,15],
          "learning_rate":[0.05,0.10,0.15,0.20,0.25,0.30],
          #"min_samples_split":[1,3,5,7],
          "gamma":[0.0,0.1,0.2,0.3,0.4],
          "colsample_bytree":[0.3,0.4,0.5,0.6]}


rf_parms={"n_estimators": np.arange(100,1500,100),
         "max_depth": np.arange(1,20),
          
         "max_features":['auto','sqrt','log2']}

In [16]:
from sklearn.tree import DecisionTreeRegressor
from  sklearn.ensemble import RandomForestRegressor

In [17]:
# ~~~~~~~~~~~~~~~~~~~~~~~~~
# Decision Tree sklearn
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~

model_11 = DecisionTreeRegressor()

model_11.fit(X_train, Y_train)

model_11_pred = model_11.predict(X_Val)

mean_abs_error_m11 = mean_absolute_error(model_11_pred, Y_Val)

print("Mean Absolute Error of Model 11 : ", mean_abs_error_m11)

Mean Absolute Error of Model 11 :  7.384102441160977


In [18]:
# ~~~~~~~~~~~~~~~~~~~~~~~~~
# Random Forest sklearn
# ~~~~~~~~~~~~~~~~~~~~~~~~~

model_12 = RandomForestRegressor()

model_12.fit(X_train, Y_train.values.ravel())

model_12_pred = model_12.predict(X_Val)

mean_abs_error_m12 = mean_absolute_error(model_12_pred, Y_Val)

print("Mean Absolute Error of Model 12 : ", mean_abs_error_m12)

Mean Absolute Error of Model 12 :  6.48129407714451


In [19]:
from sklearn.model_selection import RandomizedSearchCV

In [20]:
random_search_rf = RandomizedSearchCV(model_12, param_distributions = rf_parms, n_iter = 5, scoring = 'roc_auc', n_jobs=-1, 
                                  cv = 5, verbose = 3)

In [21]:
# Hyperparameter Optimization
start_time = time.time()

random_search_rf.fit(X_train, Y_train.values.ravel())

end_time = time.time()

print("Time taken : {0}".format(end_time - start_time))

Fitting 5 folds for each of 5 candidates, totalling 25 fits
Time taken : 1523.069235086441


In [22]:
random_search_rf.best_estimator_

RandomForestRegressor(max_depth=18, n_estimators=1000)

In [57]:
 model_13 = RandomForestRegressor(max_depth=18, max_features='sqrt', n_estimators=1000)

In [58]:
model_13.fit(X_train, Y_train.values.ravel())

model_13_pred = model_13.predict(X_Val)

mean_abs_error_m13 = mean_absolute_error(model_13_pred, Y_Val)

print("Mean Absolute Error of Model 13 : ", mean_abs_error_m13)

Mean Absolute Error of Model 13 :  6.874662317474784


In [25]:
# ~~~~~~~~~~~~~~~~~~~~~~~~~
# XGB Regressor sklearn
# ~~~~~~~~~~~~~~~~~~~~~~~~~

model_8 = XGBRegressor()

model_8.fit(X_train, Y_train)

model_8_pred = model_8.predict(X_Val)

mean_abs_error_m8 = mean_absolute_error(model_8_pred, Y_Val)

print("Mean Absolute Error of Model 8 : ", mean_abs_error_m8)

Mean Absolute Error of Model 8 :  7.169678204052809


In [26]:
random_search = RandomizedSearchCV(model_8, param_distributions = xgb_parms, n_iter = 5, n_jobs=-1, 
                                   cv = 5, verbose = 3)

In [27]:
start_time = time.time()

random_search.fit(X_train, Y_train)

end_time = time.time()

print("Time taken : {0}".format(end_time - start_time))

Fitting 5 folds for each of 5 candidates, totalling 25 fits
Time taken : 207.86626410484314


In [28]:
random_search.best_estimator_

XGBRegressor(base_score=0.5, booster='gbtree', callbacks=None,
             colsample_bylevel=1, colsample_bynode=1, colsample_bytree=0.6,
             early_stopping_rounds=None, enable_categorical=False,
             eval_metric=None, gamma=0.1, gpu_id=-1, grow_policy='depthwise',
             importance_type=None, interaction_constraints='',
             learning_rate=0.1, max_bin=256, max_cat_to_onehot=4,
             max_delta_step=0, max_depth=12, max_leaves=0, min_child_weight=1,
             missing=nan, monotone_constraints='()', n_estimators=100, n_jobs=0,
             num_parallel_tree=1, predictor='auto', random_state=0, reg_alpha=0,
             reg_lambda=1, ...)

In [29]:
random_search.best_params_

{'max_depth': 12, 'learning_rate': 0.1, 'gamma': 0.1, 'colsample_bytree': 0.6}

In [30]:
model_14 = XGBRegressor(**random_search.best_params_)


In [31]:
model_14.fit(X_train, Y_train)

model_14_pred = model_14.predict(X_Val)

mean_abs_error_m14 = mean_absolute_error(model_14_pred, Y_Val)

print("Mean Absolute Error of Model 14: ", mean_abs_error_m14)

Mean Absolute Error of Model 14:  6.397199640709845


### WITHOUT OUTLIER

In [32]:
 #treatment of the outlier
from scipy import stats
import numpy as np
train1=train.copy()
print("shape of data before outliers :",train1.shape)
# using z score to remove the outier
z=np.abs(stats.zscore(train))
W_train=train[(z<3).all(axis=1)]
print("shape of data after outlier:",W_train.shape)


W_train.reset_index(inplace=True)

shape of data before outliers : (14700, 46)
shape of data after outlier: (11163, 46)


In [33]:
X=W_train.iloc[:,0: -1]
Y =W_train.iloc[:, -1]
test_size=0.20
seed=42
X_Wtrain, X_Wval, Y_Wtrain, Y_Wval = train_test_split(X, Y, test_size=test_size,random_state=seed)

In [34]:
def Models(model, name, d):
    print("Working on {} model".format(name))
    
    
    cla = model
    cla.fit(X_Wtrain, Y_Wtrain)
    
   
    
    predW = cla.predict(X_Wval)
   
    
    MAE=mean_absolute_error(Y_Wval,predW)
    MSE=mean_squared_error(Y_Wval,predW)
    RMSE=mean_squared_error(Y_Wval,predW,squared=False)
    MAEA=mean_absolute_error(Y_Wval,predW)
    R2=r2_score(Y_Wval,predW)
    
    
    
    d['Name'].append(name)
    #d['Training ACU'].append(tr_auc)
    #d['Testing ACU'].append(te_auc)
    d["RMSE"].append(RMSE)
    d['MAE'].append(MAE)
    d['MSE'].append(MSE)
    d['MAEA'].append(MAEA)
    d['R2'].append(R2)
    
    
    
    
    print("**********"*5)
    print()
    return d
    
    
    
    
     

In [35]:

d = {'Name' : [], 'RMSE': [], 'R2': [], 'MAE': [], 'MSE': [],'MAEA': []}


models = [ [GradientBoostingRegressor(loss="squared_error",learning_rate=0.2,n_estimators=1000), 'GradientBoostRegressor'],
          [ElasticNet(max_iter=2000,normalize=True), 'ElasticNet'], 
          [LGBMRegressor(), 'LightGBM'],
          [SGDRegressor(l1_ratio=0.5,), "SGDRegressor"],

          [CatBoostRegressor(), 'CatBoost'], 
          [KernelRidge(alpha=1.0), 'kernal Ridge'],
          [LinearRegression(), 'Linear Regression'], 
          [SVR(C=1.0,epsilon=0.2), 'Support Vector Ratio' ],
         [XGBRegressor(objective="reg:linear"),"XGBRegressor"],
         [BayesianRidge(),"BayesianRidge"]
         ]




for model in models:
    d = Models(model[0], model[1], d)

acu_dataW = pd.DataFrame(data = d)

Working on GradientBoostRegressor model
**************************************************

Working on ElasticNet model
**************************************************

Working on LightGBM model
**************************************************

Working on SGDRegressor model
**************************************************

Working on CatBoost model
Learning rate set to 0.057865
0:	learn: 33.5870657	total: 22ms	remaining: 22s
1:	learn: 32.3607095	total: 41.6ms	remaining: 20.7s
2:	learn: 31.3206027	total: 62.1ms	remaining: 20.6s
3:	learn: 30.3060443	total: 81.7ms	remaining: 20.3s
4:	learn: 29.3798070	total: 103ms	remaining: 20.5s
5:	learn: 28.5074057	total: 124ms	remaining: 20.6s
6:	learn: 27.7308543	total: 150ms	remaining: 21.3s
7:	learn: 26.9586987	total: 173ms	remaining: 21.4s
8:	learn: 26.1868251	total: 195ms	remaining: 21.5s
9:	learn: 25.4496233	total: 216ms	remaining: 21.4s
10:	learn: 24.8653972	total: 237ms	remaining: 21.3s
11:	learn: 24.2818237	total: 257ms	remaining: 21.2

157:	learn: 13.2059101	total: 3.55s	remaining: 18.9s
158:	learn: 13.1954725	total: 3.56s	remaining: 18.9s
159:	learn: 13.1858496	total: 3.58s	remaining: 18.8s
160:	learn: 13.1721794	total: 3.6s	remaining: 18.8s
161:	learn: 13.1510686	total: 3.62s	remaining: 18.7s
162:	learn: 13.1357612	total: 3.64s	remaining: 18.7s
163:	learn: 13.1213727	total: 3.65s	remaining: 18.6s
164:	learn: 13.1000592	total: 3.68s	remaining: 18.6s
165:	learn: 13.0821252	total: 3.71s	remaining: 18.6s
166:	learn: 13.0663006	total: 3.73s	remaining: 18.6s
167:	learn: 13.0455775	total: 3.77s	remaining: 18.6s
168:	learn: 13.0206599	total: 3.8s	remaining: 18.7s
169:	learn: 13.0038224	total: 3.83s	remaining: 18.7s
170:	learn: 12.9776479	total: 3.85s	remaining: 18.6s
171:	learn: 12.9659357	total: 3.86s	remaining: 18.6s
172:	learn: 12.9435626	total: 3.88s	remaining: 18.6s
173:	learn: 12.9335692	total: 3.9s	remaining: 18.5s
174:	learn: 12.9242434	total: 3.92s	remaining: 18.5s
175:	learn: 12.9040472	total: 3.94s	remaining: 18

322:	learn: 11.0383421	total: 6.6s	remaining: 13.8s
323:	learn: 11.0295650	total: 6.63s	remaining: 13.8s
324:	learn: 11.0199514	total: 6.66s	remaining: 13.8s
325:	learn: 11.0059156	total: 6.68s	remaining: 13.8s
326:	learn: 10.9977638	total: 6.71s	remaining: 13.8s
327:	learn: 10.9879406	total: 6.73s	remaining: 13.8s
328:	learn: 10.9778754	total: 6.75s	remaining: 13.8s
329:	learn: 10.9683382	total: 6.77s	remaining: 13.8s
330:	learn: 10.9619847	total: 6.79s	remaining: 13.7s
331:	learn: 10.9538373	total: 6.82s	remaining: 13.7s
332:	learn: 10.9416515	total: 6.84s	remaining: 13.7s
333:	learn: 10.9256264	total: 6.87s	remaining: 13.7s
334:	learn: 10.9191479	total: 6.9s	remaining: 13.7s
335:	learn: 10.9070655	total: 6.92s	remaining: 13.7s
336:	learn: 10.9009256	total: 6.95s	remaining: 13.7s
337:	learn: 10.8935200	total: 6.96s	remaining: 13.6s
338:	learn: 10.8895667	total: 6.98s	remaining: 13.6s
339:	learn: 10.8840359	total: 7s	remaining: 13.6s
340:	learn: 10.8765487	total: 7.02s	remaining: 13.6

487:	learn: 9.8387847	total: 9.66s	remaining: 10.1s
488:	learn: 9.8351092	total: 9.68s	remaining: 10.1s
489:	learn: 9.8317664	total: 9.69s	remaining: 10.1s
490:	learn: 9.8261068	total: 9.71s	remaining: 10.1s
491:	learn: 9.8230093	total: 9.73s	remaining: 10s
492:	learn: 9.8141916	total: 9.74s	remaining: 10s
493:	learn: 9.8064190	total: 9.76s	remaining: 10s
494:	learn: 9.8038036	total: 9.78s	remaining: 9.98s
495:	learn: 9.7977161	total: 9.8s	remaining: 9.95s
496:	learn: 9.7910170	total: 9.81s	remaining: 9.93s
497:	learn: 9.7850701	total: 9.83s	remaining: 9.91s
498:	learn: 9.7804851	total: 9.86s	remaining: 9.9s
499:	learn: 9.7766723	total: 9.89s	remaining: 9.89s
500:	learn: 9.7689674	total: 9.91s	remaining: 9.87s
501:	learn: 9.7626235	total: 9.93s	remaining: 9.85s
502:	learn: 9.7556849	total: 9.95s	remaining: 9.83s
503:	learn: 9.7498050	total: 9.97s	remaining: 9.81s
504:	learn: 9.7445894	total: 9.99s	remaining: 9.79s
505:	learn: 9.7368033	total: 10s	remaining: 9.77s
506:	learn: 9.7301453	

650:	learn: 9.0263356	total: 12.8s	remaining: 6.85s
651:	learn: 9.0204276	total: 12.8s	remaining: 6.83s
652:	learn: 9.0170405	total: 12.8s	remaining: 6.81s
653:	learn: 9.0118584	total: 12.8s	remaining: 6.79s
654:	learn: 9.0077291	total: 12.9s	remaining: 6.77s
655:	learn: 9.0056193	total: 12.9s	remaining: 6.75s
656:	learn: 9.0006975	total: 12.9s	remaining: 6.73s
657:	learn: 8.9973122	total: 12.9s	remaining: 6.71s
658:	learn: 8.9930724	total: 12.9s	remaining: 6.7s
659:	learn: 8.9887457	total: 13s	remaining: 6.68s
660:	learn: 8.9828181	total: 13s	remaining: 6.66s
661:	learn: 8.9799129	total: 13s	remaining: 6.64s
662:	learn: 8.9759908	total: 13s	remaining: 6.63s
663:	learn: 8.9721525	total: 13.1s	remaining: 6.61s
664:	learn: 8.9682393	total: 13.1s	remaining: 6.59s
665:	learn: 8.9629403	total: 13.1s	remaining: 6.57s
666:	learn: 8.9572576	total: 13.1s	remaining: 6.55s
667:	learn: 8.9524372	total: 13.2s	remaining: 6.54s
668:	learn: 8.9475370	total: 13.2s	remaining: 6.52s
669:	learn: 8.9428233

814:	learn: 8.3968289	total: 15.8s	remaining: 3.59s
815:	learn: 8.3918125	total: 15.9s	remaining: 3.57s
816:	learn: 8.3889257	total: 15.9s	remaining: 3.55s
817:	learn: 8.3860578	total: 15.9s	remaining: 3.53s
818:	learn: 8.3813358	total: 15.9s	remaining: 3.51s
819:	learn: 8.3795908	total: 15.9s	remaining: 3.49s
820:	learn: 8.3760874	total: 15.9s	remaining: 3.48s
821:	learn: 8.3732573	total: 16s	remaining: 3.46s
822:	learn: 8.3696340	total: 16s	remaining: 3.44s
823:	learn: 8.3661802	total: 16s	remaining: 3.42s
824:	learn: 8.3637856	total: 16s	remaining: 3.4s
825:	learn: 8.3598277	total: 16.1s	remaining: 3.38s
826:	learn: 8.3560037	total: 16.1s	remaining: 3.37s
827:	learn: 8.3545808	total: 16.1s	remaining: 3.35s
828:	learn: 8.3512785	total: 16.1s	remaining: 3.33s
829:	learn: 8.3490115	total: 16.2s	remaining: 3.31s
830:	learn: 8.3431335	total: 16.2s	remaining: 3.29s
831:	learn: 8.3390696	total: 16.2s	remaining: 3.27s
832:	learn: 8.3358985	total: 16.2s	remaining: 3.25s
833:	learn: 8.3324585

980:	learn: 7.8696409	total: 18.9s	remaining: 367ms
981:	learn: 7.8675866	total: 18.9s	remaining: 347ms
982:	learn: 7.8647723	total: 19s	remaining: 328ms
983:	learn: 7.8616189	total: 19s	remaining: 309ms
984:	learn: 7.8583117	total: 19s	remaining: 289ms
985:	learn: 7.8557483	total: 19s	remaining: 270ms
986:	learn: 7.8533151	total: 19s	remaining: 251ms
987:	learn: 7.8506079	total: 19s	remaining: 231ms
988:	learn: 7.8477487	total: 19.1s	remaining: 212ms
989:	learn: 7.8442896	total: 19.1s	remaining: 193ms
990:	learn: 7.8427345	total: 19.1s	remaining: 173ms
991:	learn: 7.8406740	total: 19.1s	remaining: 154ms
992:	learn: 7.8374898	total: 19.1s	remaining: 135ms
993:	learn: 7.8342498	total: 19.1s	remaining: 116ms
994:	learn: 7.8308999	total: 19.2s	remaining: 96.3ms
995:	learn: 7.8284235	total: 19.2s	remaining: 77ms
996:	learn: 7.8251637	total: 19.2s	remaining: 57.8ms
997:	learn: 7.8215114	total: 19.2s	remaining: 38.5ms
998:	learn: 7.8177945	total: 19.2s	remaining: 19.3ms
999:	learn: 7.8157903

In [36]:
acu_dataW

Name          RMSE            R2           MAE  \
0  GradientBoostRegressor  1.281773e+01  8.590667e-01  8.399167e+00   
1              ElasticNet  3.414713e+01 -2.298290e-04  3.000642e+01   
2                LightGBM  1.247237e+01  8.665589e-01  8.142651e+00   
3            SGDRegressor  3.361586e+15 -9.693478e+27  2.481610e+15   
4                CatBoost  1.205786e+01  8.752812e-01  7.765926e+00   
5            kernal Ridge  2.554941e+01  4.400453e-01  2.053828e+01   
6       Linear Regression  1.922613e+01  6.829155e-01  1.494102e+01   
7    Support Vector Ratio  2.448739e+01  4.856292e-01  1.907642e+01   
8            XGBRegressor  1.251385e+01  8.656700e-01  8.026464e+00   
9           BayesianRidge  1.922302e+01  6.830179e-01  1.494078e+01   

            MSE          MAEA  
0  1.642943e+02  8.399167e+00  
1  1.166027e+03  3.000642e+01  
2  1.555601e+02  8.142651e+00  
3  1.130026e+31  2.481610e+15  
4  1.453920e+02  7.765926e+00  
5  6.527722e+02  2.053828e+01  
6  3.696441e+02  1.494102e+01  
7  5.996323e+02  1.907642e+01  
8  1.565963e+02  8.026464e+00  
9  3.695247e+02  1.494078e+01

In [38]:
model_11W = DecisionTreeRegressor()

model_11W.fit(X_Wtrain, Y_Wtrain)

model_11W_pred = model_11W.predict(X_Wval)

mean_abs_error_m11W = mean_absolute_error(model_11W_pred, Y_Wval)

print("Mean Absolute Error of Model 11 without Outliers : ", mean_abs_error_m11W)

Mean Absolute Error of Model 11 without Outliers :  8.980438871473353


In [41]:
# ~~~~~~~~~~~~~~~~~~~~~~~~~
# Random Forest sklearn
# ~~~~~~~~~~~~~~~~~~~~~~~~~

model_12W = RandomForestRegressor()

model_12W.fit(X_Wtrain, Y_Wtrain.values.ravel())

model_12W_pred = model_12W.predict(X_Wval)

mean_abs_error_m12W = mean_absolute_error(model_12W_pred, Y_Wval)

print("Mean Absolute Error of Model 12 without Outliers : ", mean_abs_error_m12W)

Mean Absolute Error of Model 12 without Outliers :  7.372010073891626


In [42]:
random_search_rf = RandomizedSearchCV(model_12W, param_distributions = rf_parms, n_iter = 5, scoring = 'roc_auc', n_jobs=-1, 
                                  cv = 5, verbose = 3)

In [43]:
# Hyperparameter Optimization for without outlier
start_time = time.time()

random_search_rf.fit(X_Wtrain, Y_Wtrain.values.ravel())

end_time = time.time()

print("Time taken for without Outlier : {0}".format(end_time - start_time))

Fitting 5 folds for each of 5 candidates, totalling 25 fits
Time taken for without Outlier : 285.0724492073059


In [44]:
random_search_rf.best_estimator_

RandomForestRegressor(max_depth=9, max_features='sqrt', n_estimators=1200)

In [45]:
 model_13W = RandomForestRegressor(max_depth=9, max_features='sqrt', n_estimators=1200)

In [47]:
model_13W.fit(X_Wtrain, Y_Wtrain.values.ravel())

model_13W_pred = model_13W.predict(X_Wval)

mean_abs_error_m13W = mean_absolute_error(model_13W_pred, Y_Wval)

print("Mean Absolute Error of Model 13 without Outlier : ", mean_abs_error_m13W)

Mean Absolute Error of Model 13 without Outlier :  10.78033086865653


In [48]:
# ~~~~~~~~~~~~~~~~~~~~~~~~~
# XGB Regressor sklearn
# ~~~~~~~~~~~~~~~~~~~~~~~~~

model_8W = XGBRegressor()

model_8W.fit(X_Wtrain, Y_Wtrain)

model_8W_pred = model_8W.predict(X_Wval)

mean_abs_error_m8W = mean_absolute_error(model_8W_pred, Y_Wval)

print("Mean Absolute Error of Model 8 without Outliers : ", mean_abs_error_m8W)

Mean Absolute Error of Model 8 without Outliers :  8.0264635324903


In [49]:
random_search = RandomizedSearchCV(model_8W, param_distributions = xgb_parms, n_iter = 5, n_jobs=-1, 
                                   cv = 5, verbose = 3)

In [50]:
start_time = time.time()

random_search.fit(X_Wtrain, Y_Wtrain)

end_time = time.time()

print("Time taken : {0}".format(end_time - start_time))

Fitting 5 folds for each of 5 candidates, totalling 25 fits
Time taken : 92.53369140625


In [51]:
random_search.best_estimator_

XGBRegressor(base_score=0.5, booster='gbtree', callbacks=None,
             colsample_bylevel=1, colsample_bynode=1, colsample_bytree=0.6,
             early_stopping_rounds=None, enable_categorical=False,
             eval_metric=None, gamma=0.3, gpu_id=-1, grow_policy='depthwise',
             importance_type=None, interaction_constraints='',
             learning_rate=0.2, max_bin=256, max_cat_to_onehot=4,
             max_delta_step=0, max_depth=12, max_leaves=0, min_child_weight=1,
             missing=nan, monotone_constraints='()', n_estimators=100, n_jobs=0,
             num_parallel_tree=1, predictor='auto', random_state=0, reg_alpha=0,
             reg_lambda=1, ...)

In [52]:
random_search.best_params_

{'max_depth': 12, 'learning_rate': 0.2, 'gamma': 0.3, 'colsample_bytree': 0.6}

In [53]:
model_14W = XGBRegressor(**random_search.best_params_)

In [56]:
model_14W.fit(X_Wtrain, Y_Wtrain)

model_14W_pred = model_14W.predict(X_Wval)

mean_abs_error_m14W = mean_absolute_error(model_14W_pred, Y_Wval)

print("Mean Absolute Error of Model 14 without Outliers: ", mean_abs_error_m14W)

Mean Absolute Error of Model 14 without Outliers:  7.46239740635788


__Minimum Mean Absolute Error -   6.397199640709845 (XGB with hyperparameter optimization with Outlier)__


# Recall

In [63]:
# make predictions on test set for 
X_test=test.values

output_prediction=model_14.predict(X_test)

In [68]:
#to check the equality
print(len(output_prediction),len(X_test))

6300 6300


In [74]:
import os
data_dir="C:/Users/Dell/Documents/Topcooder/challenge_dataset"
solution_df = pd.read_csv(os.path.join(data_dir, "test_x.csv"),header=None)
solution_df['target'] = output_prediction
solution_df.to_csv('solution.csv', index=False)